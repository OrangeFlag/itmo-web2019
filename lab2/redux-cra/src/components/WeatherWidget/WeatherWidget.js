import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import Container from "../Container/Container.style";
import Loader from "../Loader/Loader.style";

const Wait = styled.p`
  font-size: 1.5em;
  text-align: center;
  color: palevioletred;
`;


export class WeatherWidget extends React.Component {
    render() {
        const {
            weatherImg, town, temperature,
            wind, cloudiness, mercuryPressure,
            humidity, uploaded
        } = this.props;
        if (uploaded == null || uploaded === false) {
            return (
                <Container>
                    <Wait>Подождите, данные загружаются</Wait>
                    <br/>
                    <Loader/>
                </Container>)

        } else {
            return (
                <Container>
                    {town}
                    <br/>
                    <img src={weatherImg} alt=""/>
                    <br/>
                    Temperature: {temperature} °C
                    <br/>
                    Wind: {wind} m/s
                    <br/>
                    Cloudiness: {cloudiness} %
                    <br/>
                    Millimeter of mercury: {mercuryPressure} mm
                    <br/>
                    Humidity: {humidity} %
                    <br/>
                </Container>)
        }
    }
}

WeatherWidget.propTypes = {
    weatherImg: PropTypes.string,
    town: PropTypes.string,
    temperature: PropTypes.number,
    wind: PropTypes.number,
    cloudiness: PropTypes.number,
    mercuryPressure: PropTypes.number,
    humidity: PropTypes.number,
    uploaded: PropTypes.bool,
};