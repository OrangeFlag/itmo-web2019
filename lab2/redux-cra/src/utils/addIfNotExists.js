export function addIfNotExists(array, item) {
    if (array.indexOf(item) === -1){
        return [...array, item]
    } else{
        return array
    }
}