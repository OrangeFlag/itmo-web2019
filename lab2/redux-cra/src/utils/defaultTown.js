export const defaultTown = "Saint Petersburg";