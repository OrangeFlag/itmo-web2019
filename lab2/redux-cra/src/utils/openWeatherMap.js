const unsecureApiKey = "1c14e84db0be925cc69df123e5989ecc";
export const weatherAPI = `https://api.openweathermap.org/data/2.5/weather?APPID=${unsecureApiKey}`;