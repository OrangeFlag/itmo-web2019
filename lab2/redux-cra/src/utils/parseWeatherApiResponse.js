export function parseWeatherApiResponse(it) {
    return {
        town: it && it.name,
        temperature: it && it.main && it.main.temp,
        wind: it && it.wind && it.wind.speed,
        cloudiness: it && it.clouds && it.clouds.all,
        mercuryPressure: it && it.main && it.main.pressure,
        humidity: it && it.main && it.main.humidity,
    }
}