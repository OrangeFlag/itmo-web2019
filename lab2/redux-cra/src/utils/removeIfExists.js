export function removeIfExists(array, item) {
    return array.filter(it => it !== item)
}