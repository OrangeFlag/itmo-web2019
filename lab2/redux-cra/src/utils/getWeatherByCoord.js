import {weatherAPI} from "./openWeatherMap";

export async function getWeatherByCoord(latitude, longitude) {

    const response = await fetch(weatherAPI + `&lat=${latitude}&lon=${longitude}&units=metric`);

    const json = await response.json();

    return json;
}
