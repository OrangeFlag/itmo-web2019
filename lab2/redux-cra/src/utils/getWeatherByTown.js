import {weatherAPI} from "./openWeatherMap";

export async function getWeatherByTown(town) {
    const response = await fetch(weatherAPI + `&q=${town}&units=metric`);

    if (response.status !== 200){
        throw new Error('Status not 200');
    }

    return await response.json();
}
