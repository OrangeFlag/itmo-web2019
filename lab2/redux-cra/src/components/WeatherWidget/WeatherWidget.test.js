import React from 'react';
import {mount} from 'enzyme';
import {WeatherWidget} from './WeatherWidget'
import {statusEnum} from "../../utils/statusEnum";


jest.useFakeTimers();

describe('WeatherWidget test', () => {
    it('should render correctly when uploaded', () => {
        const component = mount(<WeatherWidget icon={"04d"} town={"Moscow"} temperature={20} wind={10} cloudiness={10}
                                               mercuryPressure={5} humidity={8} status={statusEnum.ok}/>);

        expect(component).toMatchSnapshot();
        component.unmount();
    });

    it('should render correctly when loading', () => {
        const component = mount(<WeatherWidget icon={"04d"} town={"Moscow"} temperature={20} wind={10} cloudiness={10}
                                               mercuryPressure={5} humidity={8} status={statusEnum.loading}/>);

        expect(component).toMatchSnapshot();
        component.unmount();
    });

    it('should render correctly when error', () => {
        const component = mount(<WeatherWidget icon={"04d"} town={"mainTown"} temperature={20} wind={10} cloudiness={10}
                                               mercuryPressure={5} humidity={8} status={statusEnum.error}/>);

        expect(component).toMatchSnapshot();
        component.unmount();
    });


    it('function should call after 5000 ms', () => {
        const clickFn = jest.fn();
        const component = mount(<WeatherWidget icon={"04d"} town={"Moscow"} temperature={20} wind={10} cloudiness={10}
                                               mercuryPressure={5} humidity={8} status={statusEnum.error}
                                               deleteFavoriteTownAction={clickFn}/>);

        jest.runAllTimers();


        expect(clickFn).toHaveBeenCalled();
        component.unmount();
    });

    it('function should not call on mainTown', () => {
        const clickFn = jest.fn();
        const component = mount(<WeatherWidget icon={"04d"} town={"mainTown"} temperature={20} wind={10}
                                               cloudiness={10}
                                               mercuryPressure={5} humidity={8} status={statusEnum.error}
                                               deleteFavoriteTownAction={clickFn}/>);

        jest.runAllTimers();


        expect(clickFn).not.toHaveBeenCalled();
        component.unmount();
    });
});