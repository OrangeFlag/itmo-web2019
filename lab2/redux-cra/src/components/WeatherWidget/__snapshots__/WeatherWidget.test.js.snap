// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeatherWidget test should render correctly when error 1`] = `
<WeatherWidget
  cloudiness={10}
  humidity={8}
  icon="04d"
  mercuryPressure={5}
  status="error"
  temperature={20}
  town="mainTown"
  wind={10}
>
  <styled.div>
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-bdVaJa",
            "isStatic": false,
            "lastClassName": "bwaJXF",
            "rules": Array [
              "
  display: flex;
  flex-flow: row wrap;
  align-items: center;

  width: 76%;

  background-color: #0138a5;
  background-color: ___CSS_1___;

  color: #ffffff;
  color: ___CSS_2___;
  
  padding: 12px 10%;
  padding-top: ___CSS_3___;
  padding-bottom: ___CSS_4___;
  
  box-shadow: 0px 2px 2px 0px rgba(0, 0, 0, 0.30);
  margin: 7px;


  * {
    margin-right: 16px;
  }
",
            ],
          },
          "displayName": "styled.div",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-bdVaJa",
          "target": "div",
          "toString": [Function],
          "usesTheme": false,
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
    >
      <div
        className="sc-bdVaJa bwaJXF"
      >
        <styled.p>
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-bxivhb",
                  "isStatic": false,
                  "lastClassName": "hYMqqw",
                  "rules": Array [
                    "
  font-size: 1.5em;
  text-align: center;
  color: #f27670;
",
                  ],
                },
                "displayName": "styled.p",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-bxivhb",
                "target": "p",
                "toString": [Function],
                "usesTheme": false,
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
          >
            <p
              className="sc-bxivhb hYMqqw"
            >
              Ошибка при запросе погоды
            </p>
          </StyledComponent>
        </styled.p>
      </div>
    </StyledComponent>
  </styled.div>
</WeatherWidget>
`;

exports[`WeatherWidget test should render correctly when loading 1`] = `
<WeatherWidget
  cloudiness={10}
  humidity={8}
  icon="04d"
  mercuryPressure={5}
  status="loading"
  temperature={20}
  town="Moscow"
  wind={10}
>
  <styled.div>
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-bdVaJa",
            "isStatic": false,
            "lastClassName": "bwaJXF",
            "rules": Array [
              "
  display: flex;
  flex-flow: row wrap;
  align-items: center;

  width: 76%;

  background-color: #0138a5;
  background-color: ___CSS_1___;

  color: #ffffff;
  color: ___CSS_2___;
  
  padding: 12px 10%;
  padding-top: ___CSS_3___;
  padding-bottom: ___CSS_4___;
  
  box-shadow: 0px 2px 2px 0px rgba(0, 0, 0, 0.30);
  margin: 7px;


  * {
    margin-right: 16px;
  }
",
            ],
          },
          "displayName": "styled.div",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-bdVaJa",
          "target": "div",
          "toString": [Function],
          "usesTheme": false,
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
    >
      <div
        className="sc-bdVaJa bwaJXF"
      >
        <styled.p>
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-htpNat",
                  "isStatic": false,
                  "lastClassName": "Fwlel",
                  "rules": Array [
                    "
  font-size: 1.5em;
  text-align: center;
  color: #e1da8a;
",
                  ],
                },
                "displayName": "styled.p",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-htpNat",
                "target": "p",
                "toString": [Function],
                "usesTheme": false,
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
          >
            <p
              className="sc-htpNat Fwlel"
            >
              Подождите, данные загружаются
            </p>
          </StyledComponent>
        </styled.p>
        <br />
        <styled.div>
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-bwzfXH",
                  "isStatic": false,
                  "lastClassName": "clutZI",
                  "rules": Array [
                    "
    border: 16px solid #f3f3f3; /* Light grey */
    border-top: 16px solid #3498db; /* Blue */
    border-radius: 50%;
    width: 120px;
    height: 120px;
    animation: ",
                    Keyframes {
                      "id": "sc-keyframes-iVXCSc",
                      "inject": [Function],
                      "name": "iVXCSc",
                      "rules": Array [
                        "@-webkit-keyframes iVXCSc{from{-webkit-transform:rotate(0deg);-ms-transform:rotate(0deg);transform:rotate(0deg);}to{-webkit-transform:rotate(360deg);-ms-transform:rotate(360deg);transform:rotate(360deg);}}",
                        "@keyframes iVXCSc{from{-webkit-transform:rotate(0deg);-ms-transform:rotate(0deg);transform:rotate(0deg);}to{-webkit-transform:rotate(360deg);-ms-transform:rotate(360deg);transform:rotate(360deg);}}",
                      ],
                      "toString": [Function],
                    },
                    " 2s linear infinite;
",
                  ],
                },
                "displayName": "styled.div",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-bwzfXH",
                "target": "div",
                "toString": [Function],
                "usesTheme": false,
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
          >
            <div
              className="sc-bwzfXH clutZI"
            />
          </StyledComponent>
        </styled.div>
      </div>
    </StyledComponent>
  </styled.div>
</WeatherWidget>
`;

exports[`WeatherWidget test should render correctly when uploaded 1`] = `
<WeatherWidget
  cloudiness={10}
  humidity={8}
  icon="04d"
  mercuryPressure={5}
  status="ok"
  temperature={20}
  town="Moscow"
  wind={10}
>
  <styled.div>
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-bdVaJa",
            "isStatic": false,
            "lastClassName": "bwaJXF",
            "rules": Array [
              "
  display: flex;
  flex-flow: row wrap;
  align-items: center;

  width: 76%;

  background-color: #0138a5;
  background-color: ___CSS_1___;

  color: #ffffff;
  color: ___CSS_2___;
  
  padding: 12px 10%;
  padding-top: ___CSS_3___;
  padding-bottom: ___CSS_4___;
  
  box-shadow: 0px 2px 2px 0px rgba(0, 0, 0, 0.30);
  margin: 7px;


  * {
    margin-right: 16px;
  }
",
            ],
          },
          "displayName": "styled.div",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-bdVaJa",
          "target": "div",
          "toString": [Function],
          "usesTheme": false,
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
    >
      <div
        className="sc-bdVaJa bwaJXF"
      >
        Moscow
        <img
          alt=""
          src="http://openweathermap.org/img/wn/04d@2x.png"
        />
        <p>
          Temperature: 
          20
           °C
          <br />
          Wind: 
          10
           m/s
          <br />
          Cloudiness: 
          10
           %
          <br />
          Millimeter of mercury: 
          5
           mm
          <br />
          Humidity: 
          8
           %
          <br />
        </p>
      </div>
    </StyledComponent>
  </styled.div>
</WeatherWidget>
`;
